### PR TITLE
Comment nodes in AST

### LINKED ISSUE #901: Comments should be AST nodes
Consider the following program:
```
from qiskit.qasm import Qasm
qasm_string = """//this is a comment
OPENQASM 2.0;
qreg q[2]; // This is another comment
"""

node_circuit = Qasm(data=qasm_string).parse()
print(node_circuit.qasm())
print([ type(i) for i in node_circuit.children])
```
The output is:
```
OPENQASM 2.0;
qreg q[2];

[<class 'qiskit.qasm._node._format.Format'>, <class 'qiskit.qasm._node._qreg.Qreg'>]
```

### What is the expected enhancement?

The following output:
```
//this is a comment
OPENQASM 2.0; 
qreg q[2];
// This is another comment

[<class 'qiskit.qasm._node._comment.Comment'>, <class 'qiskit.qasm._node._format.Format'>, <class 'qiskit.qasm._node._qreg.Qreg'>, <class 'qiskit.qasm._node._comment.Comment'>]
```

The current parser ignores comments. There is room for discussion if a parser should create an AST with this information (a Full Syntax Tree).  I think in our context is a handy thing to have. For example, to add annotations or metainformation, which might be handy in such experimental field.

### PR DESCRIPTION
**Pull request by 1ucian0**
Fixes #901.

### Summary

This PR adds support for comment nodes in AST.

### Details and comments

I added two parameters to `parse`:
 - `debug=False`: It was just convenient and it seems handy to have it. The default is `False`.
 - `with_comments`: By @delapuente suggestion in #901, the lexer considers or not comments based on this option. The default is `False`.

### CODE REVIEW

File: test/python/test_qasm_parser.py [hunk 1]
```diff
@@ -78,5 +81,37 @@ def test_get_tokens(self):
...
+    def test_single_comment(self):
+        """The thing to parse is a single comment"""
+        qasm_string = """//this is a comment"""
+        ast = Qasm(data=qasm_string).parse(with_comments=True)
+        self.assertEqual(len(ast.children), 1)
+        self.assertIsInstance(ast.children[0], Comment)
+        self.assertEqual(ast.qasm().replace('\n', '', 1), qasm_string)
+
+    def test_inline_comment(self):
+        """The thing to parse is a single comment"""
```
Comment (delapuente) [inline test/python/test_qasm_parser.py]: I think this docstring is a repetition of the previous one. Can we update it?

File: test/python/test_qasm_parser.py [hunk 2]
```diff
@@ -78,5 +81,37 @@ def test_get_tokens(self):
...
+class TestParserWithComments(QiskitTestCase):
+    """QasmParser has a with_comment option to get have comments as extra nodes"""
+
+    def test_single_comment(self):
+        """The thing to parse is a single comment"""
+        qasm_string = """//this is a comment"""
+        ast = Qasm(data=qasm_string).parse(with_comments=True)
+        self.assertEqual(len(ast.children), 1)
+        self.assertIsInstance(ast.children[0], Comment)
+        self.assertEqual(ast.qasm().replace('\n', '', 1), qasm_string)
```
Comment (delapuente) [inline test/python/test_qasm_parser.py]: Non-blocking: I'm against modifying the value I want to test, I prefer to modify the other one or provide an exact representation. Don't you think?

File: test/python/test_qasm_parser.py (hunk 1 above)
Comment (1ucian0) [reply test/python/test_qasm_parser.py]: fixed in 7068fbd

File: test/python/test_qasm_parser.py (hunk 2 above)
Comment (1ucian0) [reply test/python/test_qasm_parser.py]: I agree with the principle. The "unparser" has some issues. More details in #901.

Comment (delapuente) [reply test/python/test_qasm_parser.py]: The strings are short right now; we could, at least, write the expected string explicitely so I have not to imagine how it should look like. Is it worth the effort?

Comment (1ucian0) [reply test/python/test_qasm_parser.py]: Sure ee5afc8

**Comment by jaygambetta** [comment]
let’s put on hold and since we are going to make the qasm spec more defined let’s save this until then.

**Comment by 1ucian0** [comment]
The unparser is patched in ff1f735.

**Comment by delapuente** [comment]
Put on hold until discussing what to do with the current `QASM` module.

**Pull request closed without merging**